pkg/webhook/resources/bundledeployment/validator.go

```diff
@@ -0,0 +1,51 @@
+package bundledeployment
+
+import (
+	"fmt"
+
+	fleetvialpha1 "github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	werror "github.com/harvester/harvester/pkg/webhook/error"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewValidator() types.Validator {
+	return &bundleDeploymentValidator{}
+}
+
+type bundleDeploymentValidator struct {
+	types.DefaultValidator
+}
+
+func (v *bundleDeploymentValidator) Resource() types.Resource {
+	return types.Resource{
+		Names:      []string{fleetvialpha1.BundleDeploymentResourceName},
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   fleetvialpha1.SchemeGroupVersion.Group,
+		APIVersion: fleetvialpha1.SchemeGroupVersion.Version,
+		ObjectType: &fleetvialpha1.BundleDeployment{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Delete,
+		},
+	}
+}
+func (v *bundleDeploymentValidator) Delete(request *types.Request, oldObj runtime.Object) error {
+	bundleDeployment := oldObj.(*fleetvialpha1.BundleDeployment)
+
+	// BundleDeployment name is from Bundle.
+	// ref: https://github.com/rancher/fleet/blob/4dc66c946ca2f90f5f7a5d360a573698687a3a11/pkg/target/target.go#L370-L376
+	// Template of Bundle name is "mcc-<managedchart>".
+	// ref: https://github.com/rancher/rancher/blob/68eb5d1267679e02eb2383f0ec965b3f896ae3e3/pkg/controllers/provisioningv2/managedchart/managedchart.go#L95
+	// ManagedChart names are from:
+	// https://github.com/harvester/harvester-installer/blob/f36c8cfaa68626c85cf4c35f681dd382880f2aa7/pkg/config/templates/rancherd-10-harvester.yaml#L65-L69
+	// https://github.com/harvester/harvester-installer/blob/f36c8cfaa68626c85cf4c35f681dd382880f2aa7/pkg/config/templates/rancherd-10-harvester.yaml#L129-L133
+	if bundleDeployment != nil &&
+		(bundleDeployment.Name == "mcc-harvester" || bundleDeployment.Name == "mcc-harvester-crd") {
+		message := fmt.Sprintf("Cannot delete bundledeployment %s/%s", bundleDeployment.Namespace, bundleDeployment.Name)
+		return werror.NewInvalidError(message, "")
```

Comment:
Hmm… Is there any peril not checking namesapce here? It seems hard to determine namespace here 🥲

Reply:
Actually, I think we can get it from `fleet.cattle.io/v1alpha1/Cluster.Status.Namespace` because we already hard code target cluster name `local`.

https://github.com/harvester/harvester-installer/blob/f36c8cfaa68626c85cf4c35f681dd382880f2aa7/pkg/config/templates/rancherd-10-harvester.yaml#L77

Reply:
I add the namespace check for BundleDeployment. Please help me review it again. Thank you.